File: MetaGeta/MetaGeta.Utilities/Tuple.vb
Public Class Tuple(Of TFirst, TSecond)
    Implements IEquatable(Of Tuple(Of TFirst, TSecond))

    Private ReadOnly m_First As TFirst
    Private ReadOnly m_Second As TSecond

    Public Sub New(ByVal first As TFirst, ByVal second As TSecond)
        m_First = first
        m_Second = second
    End Sub

    Public ReadOnly Property First() As TFirst
        Get
            Return m_First
        End Get
    End Property
    Public ReadOnly Property Second() As TSecond
        Get
            Return m_Second
        End Get
    End Property

    Public Overrides Function Equals(ByVal obj As Object) As Boolean
        Return TypeOf (obj) Is Tuple(Of TFirst, TSecond) AndAlso Equals1(CType(obj, Tuple(Of TFirst, TSecond)))
    End Function
    Public Overrides Function GetHashCode() As Integer
        Return First.GetHashCode() Xor Second.GetHashCode()
    End Function
    Public Overrides Function ToString() As String
        Return String.Format("First = {0}, Second = {1}", First, Second)
    End Function

    Public Function Equals1(ByVal other As Tuple(Of TFirst, TSecond)) As Boolean Implements System.IEquatable(Of Tuple(Of TFirst, TSecond)).Equals
        Return EqualityComparer(Of TFirst).Default.Equals(First, other.First) _
        AndAlso EqualityComparer(Of TSecond).Default.Equals(Second, other.Second)
    End Function
End Class

Public Class Tuple(Of T1, T2, T3)
    Implements IEquatable(Of Tuple(Of T1, T2, T3))

    Private ReadOnly m_First As T1
    Private ReadOnly m_Second As T2
    Private ReadOnly m_Third As T3

    Public Sub New(ByVal first As T1, ByVal second As T2, ByVal third As T3)
        m_First = first
        m_Second = second
        m_Third = third
    End Sub

    Public ReadOnly Property First() As T1
        Get
            Return m_First
        End Get
    End Property
    Public ReadOnly Property Second() As T2
        Get
            Return m_Second
        End Get
    End Property

    Public ReadOnly Property Third() As T3
        Get
            Return m_Third
        End Get
    End Property

    Public Overrides Function Equals(ByVal obj As Object) As Boolean
        Return TypeOf (obj) Is Tuple(Of T1, T2, T3) AndAlso Equals1(CType(obj, Tuple(Of T1, T2, T3)))
    End Function
    Public Overrides Function GetHashCode() As Integer
        Return First.GetHashCode() Xor Second.GetHashCode() Xor Third.GetHashCode()
    End Function
    Public Overrides Function ToString() As String
        Return String.Format("First = {0}, Second = {1}, Third = {2}", First, Second, Third)
    End Function

    Public Function Equals1(ByVal other As Tuple(Of T1, T2, T3)) As Boolean Implements System.IEquatable(Of Tuple(Of T1, T2, T3)).Equals
        Return EqualityComparer(Of T1).Default.Equals(First, other.First) _
        AndAlso EqualityComparer(Of T2).Default.Equals(Second, other.Second) _
        AndAlso EqualityComparer(Of T3).Default.Equals(Third, other.Third)
    End Function
End Class

File: MetaGeta/MetaGeta.Utilities/Tuple.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections.Generic;

#endregion

namespace MetaGeta.Utilities {
    public class Tuple<TFirst, TSecond> : IEquatable<Tuple<TFirst, TSecond>> {
        private readonly TFirst m_First;

        private readonly TSecond m_Second;

        public Tuple(TFirst first, TSecond second) {
            m_First = first;
            m_Second = second;
        }

        public TFirst First {
            get { return m_First; }
        }

        public TSecond Second {
            get { return m_Second; }
        }

        #region IEquatable<Tuple<TFirst,TSecond>> Members

        bool IEquatable<Tuple<TFirst, TSecond>>.Equals(Tuple<TFirst, TSecond> other) {
            return Equals1(other);
        }

        #endregion

        public override bool Equals(object obj) {
            return (obj) is Tuple<TFirst, TSecond> && Equals1((Tuple<TFirst, TSecond>) obj);
        }

        public override int GetHashCode() {
            return First.GetHashCode() ^ Second.GetHashCode();
        }

        public override string ToString() {
            return string.Format("First = {0}, Second = {1}", First, Second);
        }

        public bool Equals1(Tuple<TFirst, TSecond> other) {
            return EqualityComparer<TFirst>.Default.Equals(First, other.First) && EqualityComparer<TSecond>.Default.Equals(Second, other.Second);
        }
    }
}

namespace MetaGeta.Utilities {
    public class Tuple<T1, T2, T3> : IEquatable<Tuple<T1, T2, T3>> {
        private readonly T1 m_First;
        private readonly T2 m_Second;

        private readonly T3 m_Third;

        public Tuple(T1 first, T2 second, T3 third) {
            m_First = first;
            m_Second = second;
            m_Third = third;
        }

        public T1 First {
            get { return m_First; }
        }

        public T2 Second {
            get { return m_Second; }
        }

        public T3 Third {
            get { return m_Third; }
        }

        #region IEquatable<Tuple<T1,T2,T3>> Members

        bool IEquatable<Tuple<T1, T2, T3>>.Equals(Tuple<T1, T2, T3> other) {
            return Equals1(other);
        }

        #endregion

        public override bool Equals(object obj) {
            return (obj) is Tuple<T1, T2, T3> && Equals1((Tuple<T1, T2, T3>) obj);
        }

        public override int GetHashCode() {
            return First.GetHashCode() ^ Second.GetHashCode() ^ Third.GetHashCode();
        }

        public override string ToString() {
            return string.Format("First = {0}, Second = {1}, Third = {2}", First, Second, Third);
        }

        public bool Equals1(Tuple<T1, T2, T3> other) {
            return EqualityComparer<T1>.Default.Equals(First, other.First) && EqualityComparer<T2>.Default.Equals(Second, other.Second) &&
                   EqualityComparer<T3>.Default.Equals(Third, other.Third);
        }
    }
}